sam.c

```diff
@@ -352,6 +364,49 @@ int32_t bam_endpos(const bam1_t *b)
         return b->core.pos + 1;
 }
 
+static int bam_tag2cigar(bam1_t *b, int recal_bin, int give_warning) // return 0 if CIGAR is untouched; 1 if CIGAR is updated with CG
+{
+    bam1_core_t *c = &b->core;
+    uint32_t cigar_st, n_cigar4, CG_st, CG_en, ori_len = b->l_data, *cigar0, CG_len, fake_bytes;
+    uint8_t *CG;
+
+    // test where there is a real CIGAR in the CG tag to move
```

Comment:
Hi @lh3, I'm adding support for cigars in CG tags into [BioJulia/BioAlignments.jl](https://github.com/BioJulia/BioAlignments.jl), and am currently implementing the test for the fake cigar, which is defined as `[k<<4|4,m<<4|3]` in the spec, where k is l_seq and m is the length of the reference in the alignment. I don't see a field I can say obviously corresponds to m in BAM records.

In your C of bam_tag2cigar, if I understand it correctly, I see you check whether the first op is a softclip, and if it's length == l_seq `if (bam_cigar_op(cigar0[0]) != BAM_CSOFT_CLIP || bam_cigar_oplen(cigar0[0]) != c->l_qseq) return 0;`
But I don't see in the C where the second cigar op of "mN" is checked, is it therefore something that is a given if kS is present then mN will be present?

Thanks,
Ben.

Reply:
There isn't a field directly in the BAM structure to define 'm' as it is the number of *reference* bases spanned by the sequence and not *query* (so not l_qseq).  The corresponding code for writing it is also in sam.c:

```
        cigar[0] = (uint32_t)c->l_qseq << 4 | BAM_CSOFT_CLIP;
        cigar[1] = (uint32_t)bam_cigar2rlen(c->n_cigar, bam_get_cigar(b)) << 4 | BAM_CREF_SKIP;
```

You'll see the bam_cigar2rlen function is the appropriate code for working out 'm'.

Reply:
@jkbonfield Ah I see, so if I'm properly understanding, since `m` is computed with `bam_cigar2rlen`, using the actual long cigar, then when reading in a BAM record, in bam_tag2cigar, only the first cigar op is checked for `k` and `BAM_CSOFT_CLIP` here:

```
// test where there is a real CIGAR in the CG tag to move
    if (c->n_cigar == 0 || c->tid < 0 || c->pos < 0) return 0;
    cigar0 = bam_get_cigar(b);
    if (bam_cigar_op(cigar0[0]) != BAM_CSOFT_CLIP || bam_cigar_oplen(cigar0[0]) != c->l_qseq) return 0;
    fake_bytes = c->n_cigar * 4;
```

Preciely because you can't yet compute `m` as the real cigar has not been read yet in order to do so?

Reply:
That would be the generous interpretation, but it probably boils down to not bothering to check. :-)  Note it's not solely entire-soft-clip, but also the presence of the CG tag.  The two combined are pretty specific, although it's not a bad idea to also validate that CG and cigar are consistent with each other.

Note the main reason for the existence of the m part of kSmN was so that tools that wanted to do coverage graphs could still use the minimal cigar string to derive the alignment end coordinate.  The code started with just kS and so the htslib implementation probably shows some of that history in how it is structured.